### DISCUSSION TITLE
Custom alert vibration patterns

### DISCUSSION THREAD

**Discussion by Navid200** [NightscoutFoundation/xDrip#3513] [Ideas]
Hello @jakobheyman,

I would like to add the option of customizing the vibration pattern for the alerts in the official xDrip.
I think this can be helpful for those who are deaf or hard of hearing.

Would you be open to collaborating on this to help us improve the official xDrip?

Thanks,

Navid

**Comment by jakobheyman** [comment]
Yes - I would happily try to help adding an option to use custom vibration patterns for the alerts in the official xDrip. In a fork of xDrip I have added this in 2021 with these two commits:
https://github.com/NightscoutFoundation/xDrip/commit/b5bec1b7a010d45fb04f9dc90c69f63fd8830e0c
https://github.com/NightscoutFoundation/xDrip/commit/7fafe286a0409ae0d1077d29565eca5d22c7ffa9

This gives the option to add a custom text string that defines the vibration pattern and it looks like this in the Edit Alert preference page:
![Screenshot_20240603-215510](https://github.com/NightscoutFoundation/xDrip/assets/9973815/701c69b4-9eb3-4c43-91e1-90764e12bd9e)

The changes are done based on trial and error and I imagine that it could be done better by someone who knows java more properly (I only know what I have learned from some changes of the xDrip code). It has worked good for me though using two different phones, Samsung A41 and Sony Xperia 10 IV. On the Xperia phone I cannot use as long vibration patterns as I could on the Samsung phone, but long enough for not being annoyed by the limit.

**Reply by Navid200** [reply]
I'm not the person who is currently leading the xDrip development.  So, everything I state is my personal opinion.
You can always open a PR ignoring my suggestions and for all I know, your PR may be approved.

My opinion:
Some of xDrip menus are too crowded.  I hope that we can clean up some of them.
My concern with crowded menus is that users can find them too complicated.
The alert edit menu is one of them.

I really really really love what you have done for implementing the custom vibration patterns.

If I have a say, I suggest this menu instead:
![Screenshot_20240603-165750](https://github.com/NightscoutFoundation/xDrip/assets/51497406/75f40a28-e6cf-4dd6-a94c-cdbef4810e4b)
  
I'm thinking of using the star character `*` to represent the cog.  Then, we won't need to worry about translation.
Tapping on that button should bring up a menu for entering the vibration pattern.  This way, we will not add anything to the page other than a star.  Everything else will go into a subpage, which the user will not need to see unless they tap on the star button.

If user enters nothing in that menu, the  default xDrip pattern should be used.

The menu should contain some description, in addition to the field for entering the pattern, to guide the user.
Since it is on a subpage, we can have a detail description explaining to the user how the string they enter corresponds to the vibration pattern they experience.

Since all patterns start with `0,`, I suggest removing that from what the user must enter and adding it to any pattern that the user chooses before submitting it to the xDrip array.  

To shorten what the user has to enter, I suggest changing the unit of the filed from ms to 10ms.
So, instead of `150,110,150,110,150,110`, the user enters `15,11,15,11,15,11`.
So, again, we will multiply everything by 10 before submitting the results to the xDrip array.

Finally, I would add some boundaries.
I would reject an entry if it is too short or too long.  We can talk about the boundaries to agree  on.
I would also put a limit on how quiet a pattern can be.  I mean the ratio of vibration time to quiet time.  Imagine a 10-second patten that only vibrates for 10ms and is quiet for 9,990ms.  That really will make no sense.  So, again we can talk about what should be acceptable.  But, it will be like the minimum acceptable ratio.

I know that this may be too much.  
I am happy to do all or some of the work to add all of this to your implementation.

What do you think?

**Reply by jakobheyman** [reply]
Thanks for your comments!
I think all your suggestions sounds very good. When I did the changes I initially intended to at least write some note about it in the xDrip discussions to offer to help get this into the official xDrip. I think the reason that I did not do so was that I felt that my code is probably ugly and that I don't know how to do it better. For example, it seemed a bit ugly to just save the pattern as a string that then has to be fixed to a long array each time it should be used. Looking at the code now, I also think that the variable vibPatLong in EditAlertActivity.java is never used and can be removed.

**Reply by Navid200** [reply]
I'm not a programmer myself.  
I have done a lot of bad things.  But, @jamorham has been extremely patient.

If we submit something that is not good, it will not be approved.

Looking at the functionality you have accomplished, I consider you to be very clever and I would love it if you join us.

**Reply by Navid200** [reply]
The most important question one needs to be able to answer when opening a PR is "Why do we need this"?
If that is reasonable, how it is implemented can be improved if needed. 

Another important factor is how could this affect users.  There are those who open a PR and when  questioned about that, say let the user decide if they want to use it or not.  If it hurts them, it is their fault.
That's not good enough.
As an example, I personally can benefit from this new alert: https://github.com/NightscoutFoundation/xDrip/pull/441
But, I am also worried about how it could negatively affect a user.  That worry has made me question the PR.  Sometimes, it is not easy to choose a patch forward!

Then, comes maintenance.  We should do everything we can to follow all the accepted standard and good practices and avoid complicating the code unnecessarily.  

Another important aspect is what xDrip is today.  If users have got used to something, it may not be a good idea to make a major change if it could confuse them.
So, the response "this will cause confusion" can be the reason for a PR not to pass.

Going back to vibration patterns, we need to think about what this change will do to the forecast low or persistent high or rising fast etc. alerts.  
Adding the ability to customize the vibration pattern for the glucose level alerts should not change the vibration pattern for those other alerts.

**Reply by jakobheyman** [reply]
Thanks for your positive comments!
I agree that it is good to be careful with changes so that they don't affect users negatively. Regarding forecast low or persistent high I would guess that they should not be affected by the changes I have done but I don't know as I have never used these alerts. I will see if I can test it in my modified version of xDrip. I imagine that there could be users who would like to be able to control such alerts as detailed as the glucose level alerts but I guess that is another question.

**Reply by Navid200** [reply]
Yes, that is.

I have wanted all alerts to become consistent:
https://github.com/NightscoutFoundation/xDrip/issues/1762
But, that is not going to be easy.  At least, not for me.

So, let's add custom vibration pattern to glucose level alerts as you have.  We can always move forward one step at a time.

The alerts are very tricky and we have to do lots of testing to show that nothing is going to break.

Thanks
